drivers: wifi: siwx91x: Add support for BG Scan and Roaming

Nitin-Pandey-01 opened the pull request:
- Added support for background scan
- Resolved the dwell time related bugs
- Updated the case for rejecting scan in channel 12, 13 and 14

muahmed-silabs commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -269,7 +304,8 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
 
-		if (z_scan_config->band_chan[i].band == WIFI_FREQ_BAND_2_4_GHZ) {
+		if (z_scan_config->band_chan[i].band == WIFI_FREQ_BAND_2_4_GHZ &&
+			z_scan_config->band_chan[i].channel <= 11) {
```
Replace with a macro

Nitin-Pandey-01 replied:
Done

NiteshKShah commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -45,6 +45,11 @@ int siwg91x_get_nwp_config(int wifi_oper_mode, sl_wifi_device_configuration_t *g
 	if (wifi_oper_mode == SL_SI91X_CLIENT_MODE) {
 		boot_config->oper_mode = SL_SI91X_CLIENT_MODE;
 
+		if (IS_ENABLED(CONFIG_WIFI_MGMT_ROAM_ENABLE)) {
```
why this ?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.h:
```diff
@@ -14,6 +14,21 @@
+#define SL_WIFI_MIN_ACTIVE_SCAN_TIME_MS  5
+#define SL_WIFI_MAX_ACTIVE_SCAN_TIME_MS  1000
+#define SL_WIFI_MIN_PASSIVE_SCAN_TIME_MS 10
+#define SL_WIFI_MAX_PASSIVE_SCAN_TIME_MS 1000
```
I don't think it it is useful to expose these values.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -432,17 +433,84 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
+					    SL_WIFI_MAX_ACTIVE_SCAN_TIME_MS) {
+					LOG_ERR("Invalid active scan dwell time");
+					return -EINVAL;
+				}
```
This function starts to be too complex (5 levels of indentation, 150 lines).

I believe you can place the input validation for dwell interval with the other ones on the top the function:
```
if (z_scan_config->dwell_time_active &&
    (z_scan_config->dwell_time_active < SL_WIFI_MIN_ACTIVE_SCAN_TIME_MS ||
     z_scan_config->dwell_time_active > SL_WIFI_MAX_ACTIVE_SCAN_TIME_MS)) {
        return -EINVAL;
}
```

You may also introduce one function to take care of `sl_wifi_set_advanced_scan_configuration()` and another one to take care of `sl_si91x_configure_timeout()`.

Note, I understand `SL_WIFI_(MAX|MIN)_(ACTIVE|PASSIVE)_SCAN_TIME_MS` are there to avoid magic value in the code. However, since they are used only once in the code and their names do not help to understand from where these value are coming from, I think you could use raw numbers here (same comment for the default values for `advanced_scan_config`).

Nitin-Pandey-01 replied:
Got it, Updated the code as suggested

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -432,17 +433,84 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
-		sl_scan_config.type = SL_WIFI_SCAN_TYPE_ACTIVE;
-		ret = sl_si91x_configure_timeout(SL_SI91X_CHANNEL_ACTIVE_SCAN_TIMEOUT,
-						 z_scan_config->dwell_time_active);
+	if (sidev->state == WIFI_STATE_COMPLETED) {
```
I wonder if it wouldn't make sense to add a field `int period` to `z_scan_config`. We would start a background scan of `period != 0`. What do you think @jukkar ?

krish2718 replied:
There isn't a bgscan support in `struct wifi_scan_params`, and if we want to add it, we need more than just `period`, no?

jerome-pouiller replied:
Indeed, we probably also need a couple of threshold values.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -432,17 +433,84 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
+			advanced_scan_config.passive_channel_time =
+				z_scan_config->dwell_time_passive;
+		}
+		advanced_scan_config.passive_channel_time = SL_WIFI_ADV_PASSIVE_SCAN_DURATION;
```
Previous assignments to this field are probably useless.

Nitin-Pandey-01 replied:
Shifted the assignments of these fields at the declaration time.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -432,17 +433,84 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
+		advanced_scan_config.trigger_level = SL_WIFI_ADV_SCAN_THRESHOLD;
+		advanced_scan_config.trigger_level_change = SL_WIFI_ADV_RSSI_TOLERANCE_THRESHOLD;
+		advanced_scan_config.enable_multi_probe = SL_WIFI_ADV_MULTIPROBE;
+		advanced_scan_config.enable_instant_scan = SL_WIFI_ENABLE_INSTANT_SCAN;
```
You can simplify the code with by assigning these value during the declaration of `advanced_scan_config`.

Nitin-Pandey-01 replied:
Done

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -45,6 +45,40 @@ config NET_MGMT_EVENT_STACK_SIZE
 	default 10
 
+config WIFI_MGMT_ENABLE_ROAM
+	bool "Roaming enabled/disabled"
```
`Enable roaming` is less ambiguous.

Nitin-Pandey-01 replied:
Done

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -45,6 +45,40 @@ config NET_MGMT_EVENT_STACK_SIZE
 config NET_MGMT_EVENT_QUEUE_SIZE
 	default 10
 
+config WIFI_MGMT_ENABLE_ROAM
```
This option is specific to siwx91x, so you have to use the proper prefix: `WIFI_SILABS_SIWX91X_ENABLE_ROAMING`

Nitin-Pandey-01 replied:
Updated the code as suggested

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -45,6 +45,40 @@ config NET_MGMT_EVENT_STACK_SIZE
+	  improving connectivity and performance during
+	  transitions between access points.
+
+config WIFI_MGMT_ENABLE_ROAM_WITH_DEAUTH
```
Ditto: `WIFI_SILABS_SIWX91X_ROAMING_USE_DEAUTH`

Nitin-Pandey-01 replied:
Used the name suggested

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -45,6 +45,40 @@ config NET_MGMT_EVENT_STACK_SIZE
+	  Enable this to roam using deauthentication frames.
+
+config WIFI_MGMT_ROAM_TRIGGER_LEVEL_CHANGE
+	int "Default value of roam hysteresis"
```
Specify the unit.

Nitin-Pandey-01 replied:
Done

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -45,6 +45,40 @@ config NET_MGMT_EVENT_STACK_SIZE
+	  frequency but may impact connectivity in weak signal areas.
+
+config WIFI_MGMT_ROAM_TRIGGER_LEVEL
+	int "Default value of roam threshold"
```
Be consistent between the name of the variable and the description. Is it a `Trigger level`, `Threshold` or an `Hysteresis`?

Nitin-Pandey-01 replied:
Sure, Added the content more clearly.
For reference, Here:
1. Trigger level change - determines the signal difference needed to switch to a new AP. Also known as Roaming Hysteresis.
2. Trigger level - determines when to start searching for a new AP. Also known as Roaming threshold.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -462,13 +464,30 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
 		advanced_scan_config.enable_instant_scan = SL_WIFI_ENABLE_INSTANT_SCAN;
+		roam_configuration.trigger_level = SL_WIFI_NEVER_ROAM;
+
+#if CONFIG_WIFI_MGMT_ENABLE_ROAM
```
Use `IS_ENABLE()`.

Nitin-Pandey-01 replied:
Updated

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -45,6 +45,40 @@ config NET_MGMT_EVENT_STACK_SIZE
+	help
+	  Enable this option to configure roaming parameters,
+	  improving connectivity and performance during
+	  transitions between access points.
```
Explain the roaming will be enable as soon as the background scan is enabled.

Nitin-Pandey-01 replied:
Done

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -45,6 +45,40 @@ config NET_MGMT_EVENT_STACK_SIZE
+	bool "Roam with Deauth enable/disable"
+	depends on WIFI_MGMT_ENABLE_ROAM
+	help
+	  Enable this to roam using deauthentication frames.
```
Can you explain a bit how this option diverge of the usual roaming?

Nitin-Pandey-01 replied:
1. Trigger level - determines when to start searching for a new AP. Also known as Roaming threshold.
2. Trigger level change - determines the signal difference needed to switch to a new AP. Also known as Roaming Hysteresis. It tells when to actually switch to a new AP.

So, In short roam threshold specifies the signal strength at which the device begins searching for a better access point, while roam hysteresis ensures the new access point's signal is sufficiently stronger than the current one to avoid unnecessary or frequent switching.

Nitin-Pandey-01 replied:
This enables roaming using Deauthentication frames, whereas, by default, the firmware performs roaming using Null data packets

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -45,6 +45,40 @@ config NET_MGMT_EVENT_STACK_SIZE
+	depends on WIFI_MGMT_ENABLE_ROAM
+	help
+	  Sets the default roam threshold. Lower values trigger
+	  earlier roaming; higher values delay it.
```
The difference with the previous option is unclear.

Nitin-Pandey-01 replied:
1. Trigger level - determines when to start searching for a new AP. Also known as Roaming threshold.
2. Trigger level change - determines the signal difference needed to switch to a new AP. Also known as Roaming Hysteresis. It tells when to actually switch to a new AP.

So, In short roam threshold specifies the signal strength at which the device begins searching for a better access point, while roam hysteresis ensures the new access point's signal is sufficiently stronger than the current one to avoid unnecessary or frequent switching.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -462,13 +464,30 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
+		/*
+		 * Use legacy roam configuration values if enabled,
+		 * otherwise use user-defined values
+		 */
```
I don't understand this comment. I don't when legacy values are used.

Nitin-Pandey-01 replied:
Removed the comment

krish2718 reviewed (commented):
Can you please split the commit cac746015025ce319abf8f8ca1c6b8f5660f2e28 in to two? implement dwell times and adding bgscan?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.h:
```diff
@@ -14,6 +14,15 @@
+#define SL_WIFI_ADV_PASSIVE_SCAN_DURATION    20
+#define SL_WIFI_ADV_MULTIPROBE               0
+#define SL_WIFI_ADV_SCAN_PERIODICITY         10
+#define SL_WIFI_ENABLE_INSTANT_SCAN          1
```
You don't need to keep them in the header file.

In addition, I am not sure they are helpful.

Nitin-Pandey-01 replied:
Yes, they are useful because they define the parameters for background scanning, which controls how it works. As these parameters are set in application only with recommended values, so I moved them to the header file. What do you suggest ?

jerome-pouiller replied:
Without the context of the code, the user won't be able to modify these values anyway. So:
  - Either a precise documentation have to be provided (and the values can migrate to Kconfig)
  - Or the value are directly used in the code and the user have the whole context to figure out the impact.

(Note, my comment only apply because these symbols are only used once and in the context of the initialisation of a struct)

Nitin-Pandey-01 replied:
We documented the BG Scan values and shifted parameters to Kconfig file

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -432,17 +439,24 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
+		sl_scan_config.type = SL_WIFI_SCAN_TYPE_ADV_SCAN;
+		sl_scan_config.periodic_scan_interval = SL_WIFI_ADV_SCAN_PERIODICITY;
+
 	} else {
```
cosmetics: avoid blank line before closing brace anf after opening brace.

Nitin-Pandey-01 replied:
Sure

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -404,6 +404,70 @@ static unsigned int siwx91x_on_scan(sl_wifi_event_t event, sl_wifi_scan_result_t
+			if (dwell_time_active < 5 || dwell_time_active > 1000) {
+				LOG_ERR("Invalid active scan dwell time");
+				return -EINVAL;
+			}
```
You can probably factorize this code at the top of the function:
```
        if (dwell_time_active && (dwell_time_active < 5 || dwell_time_active > 1000)) {
                return -EINVAL;
        }
        if (dwell_time_passive && (dwell_time_passive < 10 || dwell_time_passive > 1000)) {
                return -EINVAL;
        }
```

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -440,6 +504,10 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
 	if (sidev->state == WIFI_STATE_COMPLETED) {
+		siwx91x_configure_scan_dwell_time(
+			SL_WIFI_SCAN_TYPE_ADV_SCAN, z_scan_config->dwell_time_active,
+			z_scan_config->dwell_time_passive, &advanced_scan_config);
```
Align to the opening parenthese:
```
		siwx91x_configure_scan_dwell_time(SL_WIFI_SCAN_TYPE_ADV_SCAN,
						  z_scan_config->dwell_time_active,
						  z_scan_config->dwell_time_passive,
						  &advanced_scan_config);
```

Nitin-Pandey-01 replied:
This is done by Clang format or else I am getting compliance errors

jerome-pouiller replied:
Probably your line was too long. Is it still the case?

Nitin-Pandey-01 replied:
No, The issue got resolved, Updated the code.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -453,10 +521,20 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
 			sl_scan_config.type = SL_WIFI_SCAN_TYPE_ACTIVE;
+			ret = siwx91x_configure_scan_dwell_time(
+				SL_WIFI_SCAN_TYPE_ACTIVE, z_scan_config->dwell_time_active,
+				z_scan_config->dwell_time_passive, NULL);
```
Align to the opening parenthesis.

Nitin-Pandey-01 replied:
This is done by Clang format or else I am getting compliance errors

jhedberg replied:
What are the errors exactly? Too long lines? Seems like some refactoring to helper functions or rethinking the function flow to avoid deep indentation would be a good idea.

Nitin-Pandey-01 replied:
The error was due to a line being too long, but it was resolved, and I updated the code as recommended.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -404,6 +404,70 @@ static unsigned int siwx91x_on_scan(sl_wifi_event_t event, sl_wifi_scan_result_t
+				LOG_ERR("Invalid passive scan dwell time");
+				return -EINVAL;
+			}
+			ret = sl_si91x_configure_timeout(SL_SI91X_CHANNEL_ACTIVE_SCAN_TIMEOUT,
```
> SL_SI91X_CHANNEL_ACTIVE_SCAN_TIMEOUT

bad copy-paste ?

Nitin-Pandey-01 replied:
Yes, Missed it. Updated the code.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -404,6 +404,70 @@ static unsigned int siwx91x_on_scan(sl_wifi_event_t event, sl_wifi_scan_result_t
+				return -EINVAL;
+			}
+			ret = sl_si91x_configure_timeout(SL_SI91X_CHANNEL_ACTIVE_SCAN_TIMEOUT,
+							 dwell_time_active);
```
Since `dwell_time_active` is now a local variable, you can do:
```
    if (!dwell_time_active)
         dwell_time_active = SL_WIFI_DEFAULT_ACTIVE_CHANNEL_SCAN_TIME;
```

Nitin-Pandey-01 replied:
Oh! Updated as suggested.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -404,6 +404,70 @@ static unsigned int siwx91x_on_scan(sl_wifi_event_t event, sl_wifi_scan_result_t
+							 dwell_time_active);
+		}
+		break;
+	case SL_WIFI_SCAN_TYPE_ADV_SCAN:
```
Assert `(scan_type != SL_WIFI_SCAN_TYPE_ADV_SCAN || advanced_scan_config)`.

Nitin-Pandey-01 replied:
Sure, Updated the code.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -479,6 +480,8 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
 		.enable_instant_scan = SL_WIFI_ENABLE_INSTANT_SCAN,
 	};
+	sl_wifi_roam_configuration_t roam_configuration = {.trigger_level = SL_WIFI_NEVER_ROAM,
+							   .trigger_level_change = 0};
```
Prefered style is:
```
	sl_wifi_roam_configuration_t roam_configuration = {
		.trigger_level = SL_WIFI_NEVER_ROAM,
		.trigger_level_change = 0
	};
```

Nitin-Pandey-01 replied:
Ok

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -508,13 +511,24 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
 			SL_WIFI_SCAN_TYPE_ADV_SCAN, z_scan_config->dwell_time_active,
 			z_scan_config->dwell_time_passive, &advanced_scan_config);
 
+#if IS_ENABLED(CONFIG_WIFI_SILABS_SIWX91X_ENABLE_ROAMING)
```
You can then use usual C condition: `if (IS_ENABLED(CONFIG_WIFI_SILABS_SIWX91X_ENABLE_ROAMING)) {`

Alternately, you can set the values during variable declaration:
```
	sl_wifi_roam_configuration_t roam_configuration = {
#ifdef CONFIG_WIFI_SILABS_SIWX91X_ENABLE_ROAMING
		.trigger_level = CONFIG_WIFI_SILABS_SIWX91X_ROAMING_TRIGGER_LEVEL;
		.trigger_level_change = CONFIG_WIFI_SILABS_SIWX91X_ROAMING_TRIGGER_LEVEL_CHANGE,
#else
		.trigger_level = SL_WIFI_NEVER_ROAM,
#endif
	};
```

Nitin-Pandey-01 replied:
Oh! It looks much better, Updated the code as suggested

jerome-pouiller commented on drivers/wifi/siwx91x/Kconfig.siwx91x:
```diff
@@ -45,6 +45,45 @@ config NET_MGMT_EVENT_STACK_SIZE
+	  and performance during transitions between access points.
+
+config WIFI_SILABS_SIWX91X_ROAMING_USE_DEAUTH
+	bool "Roam with Deauth enable/disable"
```
It a boolean, we don't need to confuse the user by specifying `enable` and `disable` in the description. I suggest: `Use deauth frames for roaming`.

For the record, one day, I got that code :) :

    // enable encryption
    disable_encryption(true);

Nitin-Pandey-01 replied:
Sure

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -404,6 +404,62 @@ static unsigned int siwx91x_on_scan(sl_wifi_event_t event, sl_wifi_scan_result_t
+				SL_WIFI_ADV_PASSIVE_SCAN_DURATION;
+		} else {
+			advanced_scan_config->passive_channel_time = dwell_time_passive;
+		}
```
Probably bit of nitpick, but I would write it:
```
		if (!dwell_time_passive) {
			dwell_time_passive = SL_WIFI_ADV_PASSIVE_SCAN_DURATION;
		}
		advanced_scan_config->passive_channel_time = dwell_time_passive;
```
Hence:
  - you don't the cut any long line
  - you save one level of indentation
  - you get the same pattern than with SL_WIFI_SCAN_TYPE_ACTIVE / SL_WIFI_SCAN_TYPE_PASSIVE

Nitin-Pandey-01 replied:
Sure, Updated the code as suggested

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -433,17 +440,22 @@ static int siwx91x_scan(const struct device *dev, struct wifi_scan_params *z_sca
+		} else {
+			sl_scan_config.type = SL_WIFI_SCAN_TYPE_PASSIVE;
+		}
+
```
A blank line before a closing brace is probably not useful.

Nitin-Pandey-01 replied:
Sure

nipandey-01 committed:
driver: wifi: siwx91x: Add support for BG Scan

- Added support for background scan
  and configured the default values
  of BG scan configuration

Signed-off-by: Nitin Pandey <nitin.pandey@silabs.com>

nipandey-01 committed:
driver: wifi: siwx91x: Implement scan dwell time

- Verified and configured scan
  dwell time for active, passive
  and advance scan

Signed-off-by: Nitin Pandey <nitin.pandey@silabs.com>

nipandey-01 committed:
driver: wifi: siwx91x: Add roaming configuration

- Defined Kconfig macros for Roam config
- Added set roam configuration API call
  after BGSCAN

Signed-off-by: Nitin Pandey <nitin.pandey@silabs.com>
